Adjacency map for the solver

Rohit-Kakodkar opened the pull request:
## Description

Please describe the changes/features in this pull request.


- [x] Reads adjacency map generated by the mesher
- [x] Creates a boost graph for adjacency map with required properties

## Issue Number

If there is an issue created for these changes, link it here

## Checklist

Please make sure to check developer documentation on specfem docs.

- [ ] I ran the code through pre-commit to check style
- [ ] **THE DOCUMENTATION BUILDS WITHOUT WARNINGS/ERRORS**
- [ ] I have added labels to the PR (see right hand side of the PR page)
- [ ] My code passes all the integration tests
- [ ] I have added sufficient unittests to test my changes
- [ ] I have added/updated documentation for the changes I am proposing
- [ ] I have updated CMakeLists to ensure my code builds
- [ ] My code builds across all platforms

Rohit-Kakodkar committed:
Solver adjacency map

- [x] Reads adjacency map generated by the mesher
- [x] Creates a boost graph for adjacency map with required properties

Rohit-Kakodkar committed:
reverted dockerfile

int-ptr-ptr reviewed (approved):
The code looks good to me.

Is there a reason why you haven't updated `CMakeLists.txt` with `Boost::graph`? I have a local `boost`, but to run with `docker` and/or `act`, `BOOST_INCLUDE_LIBRARIES` and `BOOST_LIBS` need to be modified.

Additionally, unless I disable Unity, the `build-release` job gives a separate error:
```
| In file included from <SPECFEM_REPO_DIR>/build/release/_deps/boost-src/libs/serialization/src/xml_oarchive.cpp:21,
|                  from <SPECFEM_REPO_DIR>/build/release/_deps/boost-build/libs/serialization/CMakeFiles/boost_serialization.dir/Unity/unity_4_cxx.cxx:7:
| <SPECFEM_REPO_DIR>/build/release/_deps/boost-src/libs/serialization/include/boost/archive/impl/archive_serializer_map.ipp:31:11: error: redefinition of ‘class boost::archive::detail::extra_detail::map<Archive>’
|    31 |     class map : public basic_serializer_map
|       |           ^~~
| In file included from <SPECFEM_REPO_DIR>/build/release/_deps/boost-src/libs/serialization/src/xml_iarchive.cpp:21,
|                  from <SPECFEM_REPO_DIR>/build/release/_deps/boost-build/libs/serialization/CMakeFiles/boost_serialization.dir/Unity/unity_4_cxx.cxx:5:
| <SPECFEM_REPO_DIR>/build/release/_deps/boost-src/libs/serialization/include/boost/archive/impl/archive_serializer_map.ipp:31:11: note: previous definition of ‘class boost::archive::detail::extra_detail::map<Archive>’
|    31 |     class map : public basic_serializer_map
|       |           ^~~
```

Rohit-Kakodkar commented on include/enumerations/edge.hpp:
```diff
@@ -0,0 +1,18 @@
+
+enum class type : int { strongly_conforming = 1 };
+
+enum class orientation : int {
+  top,
+  right,
+  bottom,
+  left,
+  top_left,
+  top_right,
+  bottom_right,
+  bottom_left
+};
```
Hmm yes. The main reason was we aren't describing boundaries here so that would end up being confusing. However the 2 can be merged into something like `specfem::mesh_entities` - making this change

Rohit-Kakodkar committed:
Review changes

- [x] Merge vertex and edge enums for strongly conforming
- [x] Moved orientation to mesh_entities

Rohit-Kakodkar committed:
Updated boundaries enums to mesh_entities

lsawade commented on src/mesh/dim2/adjacency_map/adjacency_map.cpp:
```diff
@@ -413,13 +413,13 @@ static inline void get_all_conforming_adjacencies__test_edge(
-std::set<std::pair<int, specfem::enums::boundaries::type> >
+std::set<std::pair<int, specfem::mesh_entity::type> >
 specfem::mesh::adjacency_map::adjacency_map<specfem::dimension::type::dim2>::
     get_all_conforming_adjacencies(
         const int ispec_start,
-        const specfem::enums::boundaries::type bdry_start) const {
+        const specfem::mesh_entity::type bdry_start) const {
   // return value: fill `adj` with the conforming adjacencies
-  std::set<std::pair<int, specfem::enums::boundaries::type> > adj;
+  std::set<std::pair<int, specfem::mesh_entity::type> > adj;
 
   // container for the boundary we are currently interested in. The first one is
   // the argument
```
Doc string is messed up

lsawade reviewed (approved):
Productive!

int-ptr-ptr commented on include/enumerations/mesh_entities.hpp:
```diff
@@ -0,0 +1,16 @@
+
+namespace specfem::mesh_entity {
+
+enum class type : int {
+  bottom = 1,
+  top = 2,
+  left = 3,
+  right = 4,
+  bottom_left = 5,
+  bottom_right = 6,
+  top_right = 7,
+  top_left = 8
+};
```
Is there a significance to this ordering? It may not be worth changing from what we already have, but I would think it to make more sense if these matched the `QUA_9` ([`NGNOD = 9`](https://specfem2d.readthedocs.io/en/latest/03_mesh_generation/#notes-about-datapar_file-parameters)) node order:
```
         4 . . . . 7 . . . . 3
         .                   .
         .         eta       .
         .         |         .
         8         9--xi     6
         .                   .
         .                   .
         .                   .
         1 . . . . 5 . . . . 2
```
that is,
```suggestion
enum class type : int {
  bottom = 5,
  top = 7,
  left = 8,
  right = 6,
  bottom_left = 1,
  bottom_right = 2,
  top_right = 3,
  top_left = 4
};
```

Rohit-Kakodkar replied:
It's historical, the fortan mesher uses this numbering when writing interfaces, boundaries. Changing here would require changing the mesher and then we'd deviate from original mesher/solver conventions
